framework/src/tinyalsa/tinyalsa.c

```diff
@@ -532,50 +533,82 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
 #else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
 #endif
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
-		return -errno;
-	} else {
-		pcm->running = 1;
+	if (!pcm->running) {
+		pcm->enqueuedBytes = 0;
+		/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		bufdesc.session = pcm->session;
+#endif
+#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
+		bufdesc.numbytes = pcm->config.period_size;
+		for (pcm->bufPtr = 0; pcm->bufPtr < buf_info.nbuffers; pcm->bufPtr++)
+#else
+		bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
+		for (pcm->bufPtr = 0; pcm->bufPtr < CONFIG_AUDIO_NUM_BUFFERS; pcm->bufPtr++)
+#endif
+		{
+			bufdesc.u.pBuffer = pcm->pBuffers[pcm->bufPtr];
+			if (nbytes <= pcm->enqueuedBytes) {
```

Comment:
We dont have to check this condition. At this point, pcm->enqueuedBytes is 0. 
We have already checked nbytes <= 0 above.

Reply:
At line 573 , pcm->enqueuedBytes is being incremented, this condition is necessary to avoid extra enqueues to the device driver.